SafeSocketHandle.Unix: ensure we can't create a new context after the handle is disposed

tmds opened the pull request:
Most Socket operations first check if the Socket isn't disposed. Once we are past
that check, it's possible we create an AsyncContext for a Socket that got disposed.

tmds commented:
I'm not aware of any reported issues for this, but it could lead to weird behaviour.

Wraith2 commented:
Do you really want to create a new closed context every time? Would it be possible to cache a single lazy instance?

tmds commented:
It's a cached instance.

Wraith2 commented:
Ah, yes, i misinterpreted the diff, my bad.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -1118,6 +1118,14 @@ public void Trace(SocketAsyncContext context, string message, [CallerMemberName]
+        {
+            Closed = new SocketAsyncContext(new SafeSocketHandle());
+            Closed.Close();
+        }
```
Adding an explicit static cctor adds overhead to other static accesses on the type.  It'd be better to do this as:
```C#
internal static readonly SocketAsyncContext Closed = CreateClosedContext();

private static SocketAsyncContext CreateClosedContext()
{
    var closed = new SocketAsyncContext(new SafeSocketHandle());
    closed.Close();
    return closed;
}
```

stephentoub commented:
> I'm not aware of any reported issues for this, but it could lead to weird behaviour.

Is there a test you could write to demonstrate such behavior if this change isn't applied?  This adds some complexity and it's not clear it's actually worthwhile.

tmds commented:
> Is there a test you could write to demonstrate such behavior if this change isn't applied? This adds some complexity and it's not clear it's actually worthwhile.

I tried to write a test and learned why it isn't causing problems. When performing `StartAsyncOperation`, first there is a call to `Register`.

https://github.com/dotnet/corefx/blob/44f4102011e45bd4ac25f1c565f333302c2c3740/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs#L742-L749

This `Register` ends up doing a pinvoke with a `SafeSocketHandle` which causes an `ObjectDisposedException` on any new context after the handle is disposed. Without that `SafeHandle` (e.g. `DangerousGetHandle`), the operation could end up in the queue indefinitely.

tmds commented:
I've addressed PR feedback.
Per https://github.com/dotnet/corefx/pull/36539#issuecomment-479166683, this isn't causing issues. imo it is not that clear when looking at the code why that is the case.

tmds commented:
@stephentoub I prefer to go for this change because it specifically handles a potential issue. Otherwise I can add some comments that help make more clear why the implementation can't cause that issue.

stephentoub commented:
I'm still not clear how this is any better than what's already there.  The whole point of SafeHandles is to provide exactly the guarantees you highlight, and we shouldn't be switching to using DangerousGetHandle without DangerousAddRef, which would provide similar guarantees (in fact, we should be going in the opposite direction: it's been on my todo list to overhaul the use of safe handles and ensure we're actually just using normal safe handle marshaling... our previous changes to not do so were, in retrospect, flawed).

tmds commented:
When we close the socket, that will unsubscribe us from epoll events.
That means, if we have a callback waiting in the AsyncContext queue, it will never get invoked.

When the SafeSocketHandle is released, we call Close on the AsyncContext.
Just after that, and before we call `close`, on another thread an async operation may be started which will create a new AsyncContext instance for the SafeSocketHandle when using the getter. That one will never get Close invoked. That is what this PR is trying to avoid.

What I learned is that the new async operation can't make it into the queue. That it involves using the SafeSocketHandle in the `context.Register();` method. When we get there, the handle is already invalid and it throws ObjectDisposedException.

> just using normal safe handle marshaling... our previous changes to not do so were, in retrospect, flawed

Does that include no longer using `DangerousGetHandle` for Socket operations?
If I remember correctly, those were added to the Unix implementation for performance.
I think the Windows implementation may also be relying on `DangerousGetHandle` for being able to stop pending operations when the Handle is Disposed. If those pending operations were using SafeHandle marshalling, they would stop the handle from getting released.

stephentoub commented:
> Does that include no longer using DangerousGetHandle for Socket operations? If I remember correctly, those were added to the Unix implementation for performance.

Yes.  But the perf impact was already very small, and we made SafeHandle marshaling faster in 3.0, which removes most of any benefit that was there, anyway, and using SafeHandles as they're intended to be used makes the code more robust.

> I think the Windows implementation may also be relying on DangerousGetHandle for being able to stop pending operations when the Handle is Disposed.

I want to fix all that.  There are other, more reliable ways to achieve that.

> What I learned is that the new async operation can't make it into the queue. That it involves using the SafeSocketHandle in the context.Register(); method. When we get there, the handle is already invalid and it throws ObjectDisposedException.

But that means the issue this PR tries to prevent can't actually happen because of the SafeHandle usage, right?

tmds commented:
> I want to fix all that. There are other, more reliable ways to achieve that.

+1

> But that means the issue this PR tries to prevent can't actually happen because of the SafeHandle usage, right?

Yes, it was not obvious why it can't happen until you asked me to write a test.
If everything was using SafeHandles, it would have been clear. Now a specific use of SafeHandle in one place is what stops it from happening.

tmds commented:
Closing in favor of https://github.com/dotnet/corefx/pull/36721.
